dotnet/src/SemanticKernel/SkillDefinition/SkillCollection.cs

```diff
@@ -35,159 +36,135 @@ public SkillCollection(ILogger? log = null)
+            this.ThrowFunctionNotAvailable(skillName, functionName);
         }
 
-        this._log.LogError("Function not available: skill:{0} function:{1}", skillName, functionName);
-        throw new KernelException(
-            KernelException.ErrorCodes.FunctionNotAvailable,
-            $"Function not available {skillName}.{functionName}");
+        return func;
     }
 
+    /// <inheritdoc/>
+    public ISKFunction GetNativeFunction(string functionName) =>
+        this.GetNativeFunction(GlobalSkill, functionName);
+
     /// <inheritdoc/>
     public ISKFunction GetNativeFunction(string skillName, string functionName)
     {
-        if (this.HasNativeFunction(skillName, functionName))
+        if (!this.TryGetFunction(skillName, functionName, out ISKFunction? func) || func.IsSemantic)
         {
-            return this._skillCollection[skillName][functionName];
+            this.ThrowFunctionNotAvailable(skillName, functionName);
         }
 
-        this._log.LogError("Function not available: skill:{0} function:{1}", skillName, functionName);
-        throw new KernelException(
-            KernelException.ErrorCodes.FunctionNotAvailable,
-            $"Function not available {skillName}.{functionName}");
-    }
-
-    /// <inheritdoc/>
-    public ISKFunction GetNativeFunction(string functionName)
-    {
-        return this.GetNativeFunction(GlobalSkill, functionName);
+        return func;
     }
 
     /// <inheritdoc/>
     public FunctionsView GetFunctionsView(bool includeSemantic = true, bool includeNative = true)
     {
         var result = new FunctionsView();
 
-        if (includeSemantic)
+        if (includeSemantic || includeNative)
         {
             foreach (var skill in this._skillCollection)
             {
                 foreach (KeyValuePair<string, ISKFunction> f in skill.Value)
                 {
-                    if (f.Value.IsSemantic) { result.AddFunction(f.Value.Describe()); }
+                    if (f.Value.IsSemantic ? includeSemantic : includeNative)
+                    {
+                        result.AddFunction(f.Value.Describe());
+                    }
                 }
             }
         }
 
-        if (!includeNative) { return result; }
+        return result;
+    }
+
+    #region private ================================================================================
+
+    private SkillCollection AddFunction(ISKFunction functionInstance)
+    {
+        Verify.NotNull(functionInstance, "The function is NULL");
+
+        ConcurrentDictionary<string, ISKFunction> skill = this._skillCollection.GetOrAdd(functionInstance.SkillName, static _ => new(StringComparer.OrdinalIgnoreCase));
+        skill[functionInstance.Name] = functionInstance;
 
-        foreach (var skill in this._skillCollection)
+        return this;
+    }
+
+    private bool TryGetFunction(string skillName, string functionName, [NotNullWhen(true)] out ISKFunction? func)
```

Comment:
Will do

Reply:
why do we need to expose this publicly?

Reply:
Done

Reply:
I agree with Shawn that the "HasXx" pattern shouldn't be exposed.  It's non-standard, and the 90% case after asking whether it has the thing is you then want the thing, at which point you should have just retrieved it in a single call rather than doing two lookups.